src/Lean/Parser/Basic.lean

```diff
@@ -627,6 +627,63 @@ partial def strLitFnAux (startPos : String.Pos) : ParserFn := fun c s =>
     else if curr == '\\' then andthenFn quotedCharFn (strLitFnAux startPos) c s
     else strLitFnAux startPos c s
 
+/-- Raw strings have the syntax `r##...#"..."#...##` with zero or more `#`'s.
+If this is a valid start to a raw string (`r##...#"`),
+returns the parser state after the `"` along with the number of `#`'s -/
+partial def rawStrLitStart (c : ParserContext) (s : ParserState) : Option (ParserState × Nat) :=
```

Comment:
You should never pass the parser state in an Option because this requires the caller to use the state nonlinearly, since it has to hold on to the original state for use in the none case. Instead you should just always return a `ParserState` and ensure that it has its position reset to the start in that case.

Reply:
Thanks, I pushed a fix to that nonlinearity.